library/src/SPSApplication.ts

```diff
@@ -1,97 +1,263 @@
-
-		this.stream.addEventListener(
-			'statsReceived',
-			({ data: { aggregatedStats } }) => {
-				if (sendStatsToServerSetting.flag) {
-					this.sendStatsToSignallingServer(aggregatedStats);
-				}
-			}
-		);
-	}
-
-	handleSignallingResponse(signallingResp: string, isError: boolean) {
-		if (isError) {
-			this.showErrorOverlay(signallingResp);
-		} else {
-			this.showLoadingOverlay(signallingResp);
-		}
-	}
-
-	enforceSpecialSignallingServerUrl() {
-		// SPS needs to build a specific signalling server url based on the application name so K8s can distinguish it
-		this.stream.setSignallingUrlBuilder(() => {
-
-			// if we have overriden the signalling server URL with a .env file use it here
-			if (WEBSOCKET_URL !== undefined ) {
-				return WEBSOCKET_URL as string;
-			}
-
-			// get the current signalling url
-			let signallingUrl = this.stream.config.getTextSettingValue(TextParameters.SignallingServerUrl);
-
-			// build the signalling URL based on the existing window location, the result should be 'domain.com/signalling/app-name'
-			signallingUrl = signallingUrl.endsWith("/") ? signallingUrl + "signalling" + window.location.pathname : signallingUrl + "/signalling" + window.location.pathname;
-
-			return signallingUrl
-		});
-	}
-
-	showLoadingOverlay(signallingResp: string) {
-		this.hideCurrentOverlay();
-		this.loadingOverlay.show();
-		this.loadingOverlay.update(signallingResp);
-
-		// disable rain animation for now as perf is too poor on mobile devices
-		// this.loadingOverlay.animate();
-
-		this.currentOverlay = this.loadingOverlay;
-	}
-
-	/**
-	 * Send Aggregated Stats to the Signaling Server
-	 * @param stats - Aggregated Stats
-	 */
-	sendStatsToSignallingServer(stats: AggregatedStats) {
-		const data = new MessageStats(stats);
-		this.stream.webSocketController.webSocket.send(data.payload());
-	}
-}
+import { Application, SettingUIFlag, UIOptions } from '@epicgames-ps/lib-pixelstreamingfrontend-ui-ue5.4';
+import { AggregatedStats, SettingFlag, TextParameters, Logger } from '@epicgames-ps/lib-pixelstreamingfrontend-ue5.4';
+import { LoadingOverlay } from './LoadingOverlay';
+import { SPSSignalling } from './SignallingExtension';
+import { MessageStats } from './Messages';
+import { v4 as uuidv4 } from 'uuid';
+
+// For local testing. Declare a websocket URL that can be imported via a .env file that will override 
+// the signalling server URL builder.
+declare var WEBSOCKET_URL: string;
+declare var ENABLE_METRICS: boolean;
+
+const SupportedStats : Record<string, string> = {
+    'video_width': 'width',
+    'video_height': 'height',
+    'video_bitrate': 'bitrate',
+    'video_dropped': 'dropped',
+    'video_packets_lost': 'packets lost',
+    'video_fps': 'fps',
+    'video_pli_count': 'count',
+    'video_keyframes': 'keyframes',
+    'video_nack_count': 'nacks',
+    'video_freeze_count': 'freeze',
+    'video_jitter': 'jitter',
+    'video_frame_count': 'frame count',
+    'audio_bitrate': 'audio bitrate'
+}

```

Comment:
Can you migrate the majority of the changes in this file about adding the stats functionality into a new class that this class has an instance of? Purely just for encapsulation purposes so we can keep the core SPSApplication.ts really light.